examples/common/imgui_ui/ui.cpp

```diff
@@ -0,0 +1,216 @@
+{
+    gUiRunning = true;
+    SDL_GLContext gl_context;
+    SDL_Window * window = nullptr;
+
+    UiInit(&gl_context, &window);
+
+    ImGuiIO & io = ImGui::GetIO();
+
+    while (gUiRunning.load())
+    {
+        SDL_Event event;
+        while (SDL_PollEvent(&event))
+        {
+            ImGui_ImplSDL2_ProcessEvent(&event);
+            if ((event.type == SDL_QUIT) ||
+                (event.type == SDL_WINDOWEVENT && event.window.event == SDL_WINDOWEVENT_CLOSE &&
+                 event.window.windowID == SDL_GetWindowID(window)))
+            {
+                gUiRunning = false;
+            }
+        }
+
+        ImGui_ImplOpenGL3_NewFrame();
+        ImGui_ImplSDL2_NewFrame();
+        ImGui::NewFrame();
+
+        ui->UpdateState();
+        ui->Render();
+
+        // rendering
+        ImGui::Render();
+        glViewport(0, 0, (int) io.DisplaySize.x, (int) io.DisplaySize.y);
+        glClearColor(0, 0, 0, 0);
+        glClear(GL_COLOR_BUFFER_BIT);
+        ImGui_ImplOpenGL3_RenderDrawData(ImGui::GetDrawData());
+        SDL_GL_SwapWindow(window);
+    }
+
+    UiShutdown(&gl_context, &window);
+
+    ChipLogProgress(AppServer, "UI thread Stopped...");
+}
+
+} // namespace
+
+void ImguiUi::RunMainLoop()
+{
+    // Guaranteed to be on the main task (no chip event loop started yet)
+    ChipLoopLoadInitialState();
+
+    // Platform event loop will be on a separate thread,
+    // while the event UI loop will be on the main thread.
+    chip::DeviceLayer::PlatformMgr().StartEventLoopTask();
+
+    // SignalSafeStopMainLoop will stop this loop below
+    // or the loop exits by itself when processing a SDL
+    // exit (generally by clicking the window close icon).
+    EventLoop(this);
+
+    // Stop the chip main loop as well. This is expected to
+    // wait for the task to finish.
+    chip::DeviceLayer::PlatformMgr().StopEventLoopTask();
+}
+
+void ImguiUi::SignalSafeStopMainLoop()
+{
+    gUiRunning = false;
+}
+
+void ImguiUi::ChipLoopStateUpdate()
+{
+    assertChipStackLockedByCurrentThread();
+    for (auto it = mWindows.begin(); it != mWindows.end(); it++)
+    {
+        (*it)->UpdateState();
+    }
+}
+
+void ImguiUi::ChipLoopLoadInitialState()
+{
+    assertChipStackLockedByCurrentThread();
+    for (auto it = mWindows.begin(); it != mWindows.end(); it++)
+    {
+        (*it)->LoadInitialState();
+    }
+}
+
+void ImguiUi::Render()
+{
+    for (auto it = mWindows.begin(); it != mWindows.end(); it++)
+    {
+        (*it)->Render();
+    }
+}
+
+void ImguiUi::ChipLoopUpdateCallback(intptr_t self)
+{
+    ImguiUi * _this = reinterpret_cast<ImguiUi *>(self);
+    _this->ChipLoopStateUpdate();
+    sem_post(&_this->mChipLoopWaitSemaphore); // notify complete
+}
+
+void ImguiUi::UpdateState()
+{
+    chip::DeviceLayer::PlatformMgr().ScheduleWork(&ChipLoopUpdateCallback, reinterpret_cast<intptr_t>(this));
+    // ensure update is done when existing
+    if (sem_trywait(&mChipLoopWaitSemaphore) != 0)
```

Comment:
I don't follow this part.  We create the semaphore with a value of 0.  sem_trywait called in that state will just return EAGAIN immediately.   So there are two cases here:

1. This code wins the race and runs before the sem_post.  Then we land in the yield() call (why?) and when sem_post runs the value will increment to 1.
2. This code loses the race and runs after sem_post.  Then we will end up with the value set to 0, fine.

But what I would have _expected_ to happen here is that we just call sem_wait (without the try) so that we are guaranteed to wait for ChipLoopUpdateCallback to run...

Am I just completely missing how this is supposed to work?

Reply:
The intent was that `ChipLoopUpdateCallback` will only run as long as PlatformMgr runs and original code was assuming that platform manager may stop, so the logic here uses trywait instead of wait to not block forever in case the UI decides to stop.

In the updated logic, I guess we will never stop main thread before the UI so just waiting will work. Updated.